File: mobile/backend/app/repositories/report_repository.py
"""
Report repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository


class ReportRepository(BaseRepository):
    """Repository for Report documents."""

    def __init__(self):
        super().__init__("reports")

    async def get_by_operation(self, operation_id: str) -> List[Dict[str, Any]]:
        """Get all reports for a specific operation."""
        return await self.query(filters=[("operation_id", "==", operation_id)])

    async def get_damage_reports(self) -> List[Dict[str, Any]]:
        """Get all reports with physical damage."""
        return await self.query(filters=[("physical_damage", "==", True)])

    async def get_filtered(
        self,
        operation_id: Optional[str] = None,
        damage_only: bool = False,
    ) -> List[Dict[str, Any]]:
        """Get reports with optional filters."""
        filters = []
        if operation_id:
            filters.append(("operation_id", "==", operation_id))
        if damage_only:
            filters.append(("physical_damage", "==", True))
        return await self.query(filters=filters if filters else None)

    async def get_by_reporter(self, user_id: str) -> List[Dict[str, Any]]:
        """Get all reports filed by a specific user."""
        return await self.query(filters=[("reported_by", "==", user_id)])

File: mobile/frontend/lib/data/repositories/report_repository.dart
import 'package:frontend/data/models/report.dart';
import 'base_repository.dart';

class ReportRepository extends BaseRepository<Report> {
  ReportRepository() : super('reports');

  @override
  Report fromMap(Map<String, dynamic> map) => Report.fromMap(map);

  @override
  Map<String, dynamic> toMap(Report entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Get reports for a specific operation.
  Future<List<Report>> getByOperation(String operationId) {
    return query(
      where: 'operation_id = ?',
      whereArgs: [operationId],
      orderBy: 'created_at DESC',
    );
  }

  /// Get reports filed by a specific user.
  Future<List<Report>> getByReporter(String userId) {
    return query(
      where: 'reported_by = ?',
      whereArgs: [userId],
      orderBy: 'created_at DESC',
    );
  }

  /// Get reports with anomalies (missing, damage, or extra).
  Future<List<Report>> getAnomalies() {
    return query(
      where: 'missing_quantity > 0 OR physical_damage = 1 OR extra_quantity > 0',
      orderBy: 'created_at DESC',
    );
  }

  /// Get all reports sorted by date.
  Future<List<Report>> getAllSorted() {
    return getAll(orderBy: 'created_at DESC');
  }

  /// Get recent reports.
  Future<List<Report>> getRecent({int limit = 20}) {
    return getAll(orderBy: 'created_at DESC', limit: limit);
  }

  /// Count anomalies by type.
  Future<Map<String, int>> countAnomalies() async {
    final db = await database;
    final result = await db.rawQuery('''
      SELECT 
        SUM(CASE WHEN missing_quantity > 0 THEN 1 ELSE 0 END) as missing,
        SUM(CASE WHEN physical_damage = 1 THEN 1 ELSE 0 END) as damaged,
        SUM(CASE WHEN extra_quantity > 0 THEN 1 ELSE 0 END) as extra
      FROM reports WHERE is_deleted = 0
    ''');
    if (result.isEmpty) return {'missing': 0, 'damaged': 0, 'extra': 0};
    return {
      'missing': (result.first['missing'] as int?) ?? 0,
      'damaged': (result.first['damaged'] as int?) ?? 0,
      'extra': (result.first['extra'] as int?) ?? 0,
    };
  }
}
